### PR TITLE
[SQL] Support for VARIANT types

### PR DESCRIPTION
**Pull request by mihaibudiu**
This PR shows how SQL VARIANT are supported by the compiler.
VARIANT is the runtime type that can be used to implement efficiently JSON operations.

We had a Rust struct called SqlValue, this PR renames it to SltSqlValue, since it's only used for testing SLT.

### CODE REVIEW

**Comment by mihaibudiu** [comment]
Will have to update this PR since the Calcite code has been changed

**Comment by mihaibudiu** [comment]
This builds all the basic operations on VARIANTS available in Calcite.
The next step is to add some JSON-specific operations.

**Comment by mihaibudiu** [comment]
@gz, please review just the Rust files

File: sql-to-dbsp-compiler/lib/sqllib/src/array.rs [hunk 1]
```diff
@@ -613,3 +613,38 @@ where
         Some(value) => map_index_N_(value, map_index),
     }
 }
+
+/////////////// Variant index
+
+// Return type is always Option<Variant>, but result is never None, always a Variant
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/array.rs:619]: you don't want them to be static methods in a impl block for variant?

```
impl Variant {
fn indexV_N(self: Self, index: Option<T>) -> {
blabla}
}

bla.indexV_N(42);
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 2]
```diff
@@ -120,69 +120,120 @@ impl<D: Fallible> rkyv::Deserialize<Box<dyn StructVariant>, D> for Box<dyn Struc
...
     //Struct(Option<Box<dyn StructVariant>>),
 }
 
-// A macro for TryInto<Option<T>> for Variant
-macro_rules! try_into_option {
-    ($variant:ident, $type:ty) => {
-        impl TryInto<Option<$type>> for Variant {
-            type Error = &'static str;
+impl Variant {
+    fn getTypeString(&self) -> String {
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:149]: clippy doesn't complain about this?
```suggestion
    fn type_string(&self) -> String {
```

this should probably return &'static str

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 3]
```diff
@@ -120,69 +120,120 @@ impl<D: Fallible> rkyv::Deserialize<Box<dyn StructVariant>, D> for Box<dyn Struc
...
+            // Variant::StringMap(_) => "MAP",
+        }
+        .to_string()
+    }
 
-            fn try_into(self) -> Result<Option<$type>, Self::Error> {
-                match self {
-                    Variant::$variant(value) => Ok(value),
-                    _ => Err(concat!("Variant is not a ", stringify!($variant))),
+    pub fn index_string(&self, index: String) -> Variant {
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:175]: maybe can be?
```suggestion
    pub fn index_string<I: AsRef<str>>(&self, index: I) -> Variant {
```

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 4]
```diff
@@ -120,69 +120,120 @@ impl<D: Fallible> rkyv::Deserialize<Box<dyn StructVariant>, D> for Box<dyn Struc
...
+        match self {
+            /*
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:177]: delete?

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 5]
```diff
@@ -120,69 +120,120 @@ impl<D: Fallible> rkyv::Deserialize<Box<dyn StructVariant>, D> for Box<dyn Struc
...
-try_into_option!(Timestamp, Timestamp);
-try_into_option!(ShortInterval, ShortInterval);
-try_into_option!(LongInterval, LongInterval);
-try_into_option!(Geometry, GeoPoint);
-try_into_option!(Array, Vec<Variant>);
-try_into_option!(Map, BTreeMap<String, Variant>);
-//try_into_option!(Struct, Box<dyn StructVariant>);
+    pub fn index(&self, index: Variant) -> Variant {
+        match self {
+            /*
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:195]: I guess you want to keep them for later?

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 6]
```diff
@@ -120,69 +120,120 @@ impl<D: Fallible> rkyv::Deserialize<Box<dyn StructVariant>, D> for Box<dyn Struc
...
+            Variant::StringMap(value) => {
+                match index {
+                    Variant::String(index) => {
+                        match value.get(index) {
+                            None => Variant::SqlNull,
+                            Some(result) => result,
+                        }
+                    }
+                    _ => Variant::SqlNull,
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:204]: so we dont need this anymore?

File: sql-to-dbsp-compiler/lib/sqllib/src/variant.rs [hunk 7]
```diff
@@ -203,13 +254,44 @@ from!(Timestamp, Timestamp);
...
+where
+    Variant: From<T>,
+{
+    fn from(vec: Vec<T>) -> Self {
+        Variant::Array(vec.into_iter().map(Variant::from).collect())
+    }
+}
+
+/*
+impl<T> From<BTreeMap<String, T>> for Variant
```
Comment (gz) [inline sql-to-dbsp-compiler/lib/sqllib/src/variant.rs:269]: this one we probably don't need though?

**Review by gz — APPROVED** [review]

File: sql-to-dbsp-compiler/lib/sqllib/src/array.rs (hunk 1 above)
Comment (mihaibudiu) [reply sql-to-dbsp-compiler/lib/sqllib/src/array.rs:619]: so far the compiler always calls functions, sometimes the functions call methods

**Pull request merged**